### PR TITLE
Fix useNumberFieldStately dep on aria

### PR DESCRIPTION
**Pull request by snowystinger**
Closes 
Make use of https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

**Review by ktabors — DISMISSED** [review]
This change worked. I only tested NumberField

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 1]
```diff
@@ -105,9 +104,17 @@ export function useNumberFieldState(
   // Update the input value when the number value or format options change. This is done
   // in a useEffect so that the controlled behavior is correct and we only update the
   // textfield after prop changes.
-  useLayoutEffect(() => {
+  let [prevValue, setPrevValue] = useState(numberValue);
+  let [prevLocale, setPrevLocale] = useState(locale);
+  let [prevFormatOptions, setPrevFormatOptions] = useState(formatOptions);
```
Comment (devongovett) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:109]: These should be refs, not state. Setting them should not trigger a re-render. You could potentially also combine them into a single ref, that's an object containing them all?

Comment (snowystinger) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts:109]: if they haven't changed, then setting then won't cause a re-render
we only get in here if one of them has changed, and we already call setState on the main one we're interested in
the setStates will get aggregated and result in only one new render

Note, this is the way that the react docs say to implement it as well https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

Comment (devongovett) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts:109]: but I still don't get why it needs to be a state and not a ref? ref seems better for all things that don't need to cause a render.

Comment (snowystinger) [reply packages/@react-stately/numberfield/src/useNumberFieldState.ts]: yeah, i'm not sure why they implement it with state, I've moved it over anyways. they should have the same behavior

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 2]
```diff
@@ -105,9 +104,17 @@ export function useNumberFieldState(
   // Update the input value when the number value or format options change. This is done
   // in a useEffect so that the controlled behavior is correct and we only update the
   // textfield after prop changes.
-  useLayoutEffect(() => {
+  let prevValue = useRef(numberValue);
+  let prevLocale = useRef(locale);
+  let prevFormatOptions = useRef(formatOptions);
+  // because NaN === NaN => false
+  let notEquivalent = isNaN(prevValue.current) || isNaN(numberValue) ? Boolean(isNaN(prevValue.current)) !== Boolean(isNaN(numberValue)) : numberValue !== prevValue.current;
```
Comment (devongovett) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:111]: You can use `Object.is` for this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is

**Review by devongovett — APPROVED** [review]

**Review by dannify — APPROVED** [review]

**Pull request merged**